### ISSUE TITLE
Getting legend for multilayer chart

### ISSUE THREAD

**Issue by ajasja** [vega/altair#984]
Is it possible to make a legend for such a chart

```python
import altair as alt
import numpy as np
import pandas as pd

x = np.arange(100)
data = pd.DataFrame({'x': x,
                     'sin(x)': np.sin(x / 5),
                     'data': np.sin(x / 5) + 0.3*np.random.rand(100)})

line = alt.Chart(data).mark_line(strokeWidth=6, color='orange').encode(
    x='x',
    y='sin(x)'
)

point = alt.Chart(data).mark_point(color='black').encode(
    x='x',
    y='data'
)
(line + point)
```
![image](https://user-images.githubusercontent.com/19926/42045721-8b773764-7afc-11e8-943f-93d60478d3de.png)

**Comment by ajasja** [comment]
I would like to get something like this
![image](https://user-images.githubusercontent.com/19926/42045842-d873e94a-7afc-11e8-83d1-26851e5b8ae4.png)

**Comment by jakevdp** [comment]
Legends are only created if the data within a layer is somehow grouped by a label. You can force this by adding columns of labels; for example:
```python
x = np.arange(100)
data = pd.DataFrame({'x': x,
                     'sin(x)': np.sin(x / 5),
                     'data': np.sin(x / 5) + 0.3*np.random.rand(100),
                     'line_label': 100 * ['line'],
                     'points_label': 100 * ['points']})

line = alt.Chart(data).mark_line(strokeWidth=6, color='orange').encode(
    x='x',
    y='sin(x)',
    opacity='line_label'
)

point = alt.Chart(data).mark_point(color='black').encode(
    x='x',
    y='data',
    shape='points_label'
)
(line + point)
```
![visualization 25](https://user-images.githubusercontent.com/781659/42046100-f5240eb2-7ab1-11e8-8c86-ef661a6ce81f.png)

though this is admittedly a bit hacky. Also, as far as I know vega is incapable of displaying line marks within a legend as you show in your example above, though @kanitw or @domoritz may be able to correct me on that.

**Comment by ajasja** [comment]
Thanks! This is indeed a bit hacky:)
But I got even a bit closer.

```python
x = np.arange(100)
data = pd.DataFrame({'x': x,
                     'sin(x)': np.sin(x / 5),
                     'data': np.sin(x / 5) + 0.3*np.random.rand(100),
                     'line_label': 100 * ['line'],
                     'points_label': 100 * ['points']})

line = alt.Chart(data).mark_line(strokeWidth=6, color='orange').encode(
    x='x',
    y='sin(x)',
    opacity=alt.Opacity('line_label', legend=alt.Legend(title=""))
)

point = alt.Chart(data).mark_point(color='black').encode(
    x='x',
    y='data',
    shape=alt.Shape('points_label', legend=alt.Legend(title=""))
)
(line + point)
```

PS: this is probably a different debate (e.g. https://github.com/altair-viz/altair/issues/947);
I just found out I could do
```python
shape=alt.Shape('points_label', title="" )
```

instead of 

```python
shape=alt.Shape('points_label', legend=alt.Legend(title="") )
```

Big kudos points! :+1:  Is this documented somewhere or is it a more try and see?

**Comment by jakevdp** [comment]
The ``title`` keyword is documented in the API docs (e.g. [``alt.X``](https://altair-viz.github.io/user_guide/API.html#altair.X) and used in a few examples, but I think it would be useful to have a section of the documentation dedicated to titles and labels.

Any volunteers? :smile:

**Comment by pletka** [comment]
Looks great! How can you sort the labels in the legend?

**Comment by domoritz** [comment]
If you don't need to use different mark types for the layers, you can also use the fold transform documented at https://vega.github.io/vega-lite/docs/fold.html to convert you data to long/tidy form.

**Comment by jetilton** [comment]
Hi folks, I am new to altair and trying to plot weather/hydrograph data.  I am able to plot the data, but I can't seem to specify the colors I want with a legend.  My data and plot looks like the below.

```
import numpy as np
import altair as alt
x = np.arange(100)
sin = np.sin(x / 5)
data = pd.DataFrame({'x': x,
                     'sin(x)': np.sin(x / 5),
                     'q_95': sin + 100*np.random.rand(100),
                     'q_75': sin + 75*np.random.rand(100),
                     'q_50': sin + 50*np.random.rand(100),
                     'q_25': sin + 25*np.random.rand(100),
                     'q_05': sin + 5*np.random.rand(100),
                    })
perc_90 = alt.Chart(data).mark_area(color='#4292c6', opacity = .5,).encode(
    x=alt.X('x',axis=alt.Axis(title='Day')),
    y=alt.Y('q_05',axis=alt.Axis(title='cfs')),
    y2 = 'q_95',
    #fill=alt.Color("p90", legend=alt.Legend(title=''))
    
).properties(
    width=800)


perc_50 = alt.Chart(data).mark_area(color='#08519c', opacity = .5).encode(
    x=alt.X('x',axis=alt.Axis(title='Day')),
    y=alt.Y('q_25',axis=alt.Axis(title='cfs')),
    y2 = 'q_75',
    #color=alt.Color("p50", legend=alt.Legend(title=''))
    
)

median = alt.Chart(data).mark_line(color = '#08306b').encode(
    x='x',
    y='q_50',
    #opacity=alt.Color("median", legend=alt.Legend(title=''))
)

perc_90 + perc_50 + median
```


Out of curiosity is there a reason why altair does not allow for custom legends?  Thanks, I really love the work so far.

**Comment by domoritz** [comment]
@jetilton Vega-Lite supports custom legends (and so does Altair). You may need to modify the scale domain and range as in https://vega.github.io/vega-lite/examples/stacked_bar_weather.html. If you have a smaller example, I can give more feedback.

**Comment by footfalcon** [comment]
@domoritz Would you mind taking a look at this small example? I am aiming to use a selector to toggle different layered time-series but can't figure out how to generate a proper legend. This  example takes the stock price dataset and I added a dummy 'Price-Earnings' ratio to layer onto the plot, and then use another single axis plot to dashboard-toggle which stock to display. The legend I want to display should identify the 'Price' and 'PE' series instead of the symbols. I understand that the data probably has to be rearranged somehow, and it may not be practical/possible, in which case, is there a way to manually create/label a legend/textbox for this use case? Thanks in advance! 

```
    #* Testing to get legend....
    stockdata = data.stocks()
    stockdata['pe'] = stockdata['price'] / 10

    selector = alt.selection_single(
        fields=['symbol'], 
        empty='all',
        init={'symbol': 'AAPL'}
    )

    legend = alt.Chart(stockdata).mark_square(size=150).encode(
        y=alt.Y(
            'symbol:N',
            axis=alt.Axis(domain=False, ticks=False, orient='right'), title=None
        ),
        color=alt.condition(selector, 'symbol:N', alt.value('gainsboro'), legend=None)
    ).add_selection(
        selector
    )

    price = alt.Chart(stockdata).mark_line(point=True).encode(
        x='date:T',
        y='price:Q',
        color='symbol:N',
        #size='pe:Q'
    )

    pe = alt.Chart(stockdata).mark_bar().encode(
        x='date:T',
        y='pe:Q',
        color='symbol:N'
    )

    legend | (price + pe).add_selection(
                            selector
                        ).transform_filter(
                            selector
                        )
```

<img width="587" alt="image" src="https://user-images.githubusercontent.com/37611166/75442506-994cbc00-599a-11ea-8fe3-e30a8788481b.png">

**Comment by jakevdp** [comment]
For one thing, it's now possible to make native legends interactive:
```python
import altair as alt
from vega_datasets import data

stockdata = data.stocks()
stockdata['pe'] = stockdata['price'] / 10

selector = alt.selection_single(
    fields=['symbol'], 
    empty='all',
    init={'symbol': 'AAPL'},
    bind='legend'
)

price = alt.Chart(stockdata).mark_line(point=True).encode(
    x='date:T',
    y='price:Q',
    color='symbol:N',
    opacity=alt.condition(selector, alt.value(1), alt.value(0))
).add_selection(
    selector
)

pe = alt.Chart(stockdata).mark_bar().encode(
    x='date:T',
    y='pe:Q',
    color='symbol:N'
).transform_filter(
    selector
)

price + pe
```
![visualization - 2020-02-27T054454 350](https://user-images.githubusercontent.com/781659/75450787-4c2e0300-5924-11ea-810c-3afbc2a64930.png)

Beyond that, it's not clear to me how you want your legend to be different than what is shown. Both layers have a shared color encoding that is correctly reflected in the legend.

**Comment by footfalcon** [comment]
Hi Jake - thanks for your reply. I'm am aware of native legend interactivity (which is great). I should have probably mentioned more that I am new to Altair and exploring its possibilities. In this case, I am trying to see where I can take it as a mini-dashboard. The reason I want to try using the selector the way I have it is that, in my use-case:  

1. it is a long list of countries (which would get truncated as a native legend), and 
2. I want it to control several more separate plots (that would all be filtered by country selection, and 
3. I want the flexibility to control the layout.

Also, while your plot is effectively the same as mine, and the native legend _does_ identify the stock correctly by color, it does not clearly show which series is the stock **price** and which is the stock **PE**. What I am hoping to do by creating the pseudo-legend is keep that stock identity, but also be able to display a legend which says the `mark_line` series is the price, and the `mark_bar` series is the **PE**.

This may not be possible, in which case, is it possible to create something like a text box to manually place on chart? I will actually be using a different color for the line and bars (which will remain constant for each stock,eg: price == red; PE == gray), so I could color code the labels in a text box to convey that information.

```
    stockdata = data.stocks()
    stockdata['pe'] = stockdata['price'] / 10

    selector = alt.selection_single(
        fields=['symbol'], 
        empty='all',
        init={'symbol': 'AAPL'}
    )

    legend = alt.Chart(stockdata).mark_square(size=150).encode(
        y=alt.Y(
            'symbol:N',
            axis=alt.Axis(domain=False, ticks=False, orient='right'), title=None
        ),
        color=alt.condition(selector, alt.value('firebrick'), alt.value('gainsboro'), legend=None)
    ).add_selection(
        selector
    )

    price = alt.Chart(stockdata).mark_line(point=True).encode(
        x='date:T',
        y='price:Q',
        color=alt.value('firebrick'),
        #size='pe:Q'
    )

    pe = alt.Chart(stockdata).mark_bar().encode(
        x='date:T',
        y='pe:Q',
        color=alt.value('gray')
    )

    legend | (price + pe).add_selection(
                            selector
                        ).transform_filter(
                            selector
                        )
```
<img width="698" alt="image" src="https://user-images.githubusercontent.com/37611166/75455553-7d561400-59b4-11ea-925b-928824c599e2.png">

**Comment by domoritz** [comment]
Unfortunately, I don't have time right now to look at anything but minimal examples that demonstrate a specific issue.

**Comment by footfalcon** [comment]
@domoritz No problem, I will keep exploring. Am really impressed with Altair!

**Comment by jakevdp** [comment]
You could do something like this:
```python
import altair as alt
from vega_datasets import data

stockdata = data.stocks()
stockdata['pe'] = stockdata['price'] / 10

selector = alt.selection_single(
    fields=['symbol'], 
    empty='all',
    init={'symbol': 'AAPL'},
    bind='legend'
)

price = alt.Chart(stockdata).mark_line(point=True).encode(
    x='date:T',
    y='price:Q',
    color='symbol:N',
    opacity=alt.condition(selector, alt.value(1), alt.value(0))
).add_selection(
    selector
)

pe = alt.Chart(stockdata).transform_calculate(
    name='"PE Ratio"'  
).mark_bar().encode(
    x='date:T',
    y='pe:Q',
    color=alt.Color('name:N', scale=alt.Scale(scheme='greys'), legend=alt.Legend(title=None))
).transform_filter(
    selector
)

(price + pe).resolve_scale(color='independent')
```
![visualization (63)](https://user-images.githubusercontent.com/781659/75475686-1271f200-594e-11ea-9c4c-df1f78fde595.png)

The grammar offers a lot of possibilities for customizing legends and scales, depending on exactly what you want to do.

**Comment by footfalcon** [comment]
Thanks, I will give it a try...

**Comment by RobbyJS** [comment]
Hello,
Since it has been some time since this question was asked, I wanted to see if there is any updates:
is there a way of doing this (adding a legend when there is only one group of data inside the graph) that doesn't involve adding a column to the data and adding an additional property to the graph?

> Is it possible to make a legend for such a chart
> 
> ```python
> import altair as alt
> import numpy as np
> import pandas as pd
> 
> x = np.arange(100)
> data = pd.DataFrame({'x': x,
>                      'sin(x)': np.sin(x / 5),
>                      'data': np.sin(x / 5) + 0.3*np.random.rand(100)})
> 
> line = alt.Chart(data).mark_line(strokeWidth=6, color='orange').encode(
>     x='x',
>     y='sin(x)'
> )
> 
> point = alt.Chart(data).mark_point(color='black').encode(
>     x='x',
>     y='data'
> )
> (line + point)
> ```
> 
> ![image](https://user-images.githubusercontent.com/19926/42045721-8b773764-7afc-11e8-943f-93d60478d3de.png)

**Comment by jakevdp** [comment]
No, there is still no way to add a legend without specifying an encoding that the legend will represent.

**Comment by essafik** [comment]
Are there any plans to have legends not based on a label? While in many cases it is easy to add a column, it not always practical as when you have simulation data involving many parameters and wanting to compare results from different simulations on the same plot.

**Comment by jakevdp** [comment]
What do you mean by “legend not based on a label”? How do you imagine specifying what the legend will contain?

**Comment by essafik** [comment]
Before switching to Altair, I was doing plots with Mathematica and you can simply specify your legend withing the plot by using the option PlotLegend ->{"line"}. But even with matplotlib you can specify the legend with the label option as in : plot(x, y, label="line"). Is something like that planned for Altair or even possible with vega-lite?

**Comment by jakevdp** [comment]
Yes, in newer versions of vega-lite you can set encodings to a constant datum value, which will be used to populate the legend. Altair doesn't yet support this, though.

**Comment by gustavz** [comment]
when will this feature be available? As for my understanding of a plotting library its crucial.

**Comment by jakevdp** [comment]
> when will this feature be available? 

What specifically are you asking about?

**Comment by dsandber** [comment]
> What specifically are you asking about?

I believe @gustavz was asking about the ability to do "color=alt.datum("My Line")".  Either way, I'd like to know also!

**Comment by jakevdp** [comment]
You can currently specify a color like `"orange"` using `color=alt.value("orange")`

**Comment by dsandber** [comment]
@jakevdp yeah, the question is once the functionality described by @gustavz is implemented, so that a legend item can be specified by doing "color=alt.datum("My Line")", then how can the color also be specified since the "color" was set to "My Line".

**Comment by jakevdp** [comment]
You can define the scale in the normal way; i.e. `scale=alt.Scale(domain=["My Line"], range=["orange"])`

**Comment by NoName115** [comment]
I would like to add my solution to this issue, as I was struggling a lot to create a "custom legend" for my charts.
My problem was that I had `Chart(data).mark_line()` and then a created `transform_loess` from that chart, where I wanted to show that one line contain exact measured values and the other is smoothed.
I used an approach from https://github.com/altair-viz/altair/issues/2430.
My result is below:

![r](https://user-images.githubusercontent.com/18234353/120458068-43ce3e80-c397-11eb-8f61-c25efd771945.png)

**Comment by cjw296** [comment]
> In Altair it would probably look something like this (Note that this does not work in the current release):

@jakevdp - what still needs to happen for this to work in a released version of Altair?

**Comment by joelostblom** [comment]
@cjw296 You can follow the discussion of updating Altair to the recent Vega-Lite versions in this issue https://github.com/altair-viz/altair/issues/2425

**Comment by dangotbanned** [comment]
Closing as I cannot find anything actionable here. 

https://github.com/vega/altair/issues/2425 led to https://github.com/vega/altair/releases/tag/v5.0.0

Happy to reopen if someone can point to what isn't covered in
- https://altair-viz.github.io/user_guide/compound_charts.html
- https://altair-viz.github.io/user_guide/customization.html#adjusting-the-legend
- https://altair-viz.github.io/user_guide/interactions/parameters.html#selection-targets
- https://altair-viz.github.io/gallery/line_chart_with_color_datum.html
- https://altair-viz.github.io/gallery/line_chart_with_custom_legend.html
- https://altair-viz.github.io/gallery/scatter_linked_table.html
- https://altair-viz.github.io/gallery/multifeature_scatter_plot.html
- https://altair-viz.github.io/gallery/scatter_with_minimap.html
- https://altair-viz.github.io/gallery/interactive_legend.html
- https://altair-viz.github.io/gallery/scatter_point_paths_hover.html
- https://altair-viz.github.io/gallery/london_tube.html
- https://altair-viz.github.io/gallery/seattle_weather_interactive.html

**Comment by amueller** [comment]
In case someone is still looking for this:
```python
import altair as alt
import numpy as np
import pandas as pd

x = np.arange(100)
data = pd.DataFrame({'x': x,
                     'sin(x)': np.sin(x / 5),
                     'data': np.sin(x / 5) + 0.3*np.random.rand(100)})

line = alt.Chart(data).mark_line(strokeWidth=6).encode(
    x='x',
    y='sin(x)',
    color=alt.datum("Fit")
)

point = alt.Chart(data).mark_point().encode(
    x='x',
    y='data',
    color=alt.datum("Experiment")
)
(line + point)
```
now yields 

<img width="488" height="347" alt="Image" src="https://github.com/user-attachments/assets/648e4334-39f7-4c7b-8232-e8191f6a7069" />

which seems pretty close.

I'd like to also control the color though (orange and black in the example above), and I couldn't figure that out. Is there a way to make that happen? Something like

```python
domain = ['Fit', 'Experiment']
range_ = ['orange', 'black']
scale = alt.Scale(domain=domain, range=range_)

x = np.arange(100)
data = pd.DataFrame({'x': x,
                     'sin(x)': np.sin(x / 5),
                     'data': np.sin(x / 5) + 0.3*np.random.rand(100)})

line = alt.Chart(data).mark_line(strokeWidth=6).encode(
    x='x',
    y='sin(x)',
    color=alt.Color(alt.datum("Fit"), scale=scale)
)

point = alt.Chart(data).mark_point().encode(
    x='x',
    y='data',
    color=alt.Color(alt.datum("Experiment"), scale=scale)
)
(line + point)
```

unfortunately that doesn't seem possible.
Is there a way to combine custom scales with alt.datum? (also hi jake in case you read this :)

**Comment by joelostblom** [comment]
I believe it should work when tweaked slightly to `color=alt.datum("Experiment", scale=scale)`, but I'm currently running into a schema validation error in Vega-Lite, which I opened an issue for here https://github.com/vega/vega-lite/issues/9675.
